Content.Server/Body/Systems/RespiratorSystem.cs

```diff
@@ -164,6 +172,112 @@ public void Exhale(EntityUid uid, BodyComponent? body = null)
+    /// gasses).
+    /// </summary>
+    public bool CanMetabolizeInhaledAir(Entity<RespiratorComponent?> ent)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        var ev = new InhaleLocationEvent();
+        RaiseLocalEvent(ent, ref ev);
+
+        var gas = ev.Gas ?? _atmosSys.GetContainingMixture(ent.Owner);
+        if (gas == null)
+            return false;
+
+        return CanMetabolizeGas(ent, gas);
+    }
+
+    /// <summary>
+    /// Check whether or not an entity can metabolize the given gas mixture without suffocating or taking damage
+    /// (i.e., no toxic gasses).
+    /// </summary>
+    public bool CanMetabolizeGas(Entity<RespiratorComponent?> ent, GasMixture gas)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        var organs = _bodySystem.GetBodyOrganComponents<LungComponent>(ent);
+        if (organs.Count == 0)
+            return false;
+
+        gas = new GasMixture(gas);
+        var lungRatio = 1.0f / organs.Count;
+        gas.Multiply(MathF.Min(lungRatio * gas.Volume/Atmospherics.BreathVolume, lungRatio));
+        var solution = _lungSystem.GasToReagent(gas);
+
+        float saturation = 0;
+        foreach (var organ in organs)
+        {
+            saturation += GetSaturation(solution, organ.Comp.Owner, out var toxic);
+            if (toxic)
+                return false;
+        }
+
+        return saturation > ent.Comp.UpdateInterval.TotalSeconds;
+    }
+
+    /// <summary>
+    /// Get the amount of saturation that would be generated if the lung were to metabolize the given solution.
+    /// </summary>
+    /// <remarks>
+    /// This assumes the metabolism rate is unbounded, which generally should be the case for lungs, otherwise we get
+    /// back to the old pulmonary edema bug.
+    /// </remarks>
+    /// <param name="solution">The reagents to metabolize</param>
+    /// <param name="lung">The entity doing the metabolizing</param>
+    /// <param name="toxic">Whether or not any of the reagents would deal damage to the entity</param>
+    private float GetSaturation(Solution solution, Entity<MetabolizerComponent?> lung, out bool toxic)
+    {
+        toxic = false;
+        if (!Resolve(lung, ref lung.Comp))
+            return 0;
+
+        if (lung.Comp.MetabolismGroups == null)
+            return 0;
+
+        float saturation = 0;
+        foreach (var (id, quantity) in solution.Contents)
+        {
+            var reagent = _protoMan.Index<ReagentPrototype>(id.Prototype);
+            if (reagent.Metabolisms == null)
+                continue;
+
+            if (!reagent.Metabolisms.TryGetValue(GasId, out var entry))
+                continue;
+
+            foreach (var effect in entry.Effects)
+            {
+                if (effect is HealthChange health)
+                    toxic |= CanMetabolize(health) && health.Damage.Any();
```

Comment:
should this line not check if the damage is positive and not just any value? You could theoretically have healing gas

Reply:
Its a poorly named method, ill rename it to `AnyPositive()` cause that's what its actually doing